pykwasm/src/pykwasm/run_wasm.py

```diff
@@ -0,0 +1,557 @@
+"""
+
+from __future__ import annotations
+
+from enum import Enum
+from io import BytesIO, StringIO
+import os
+from pathlib import Path
+import subprocess
+import sys
+from typing import TYPE_CHECKING
+
+from wasm import instructions
+from wasm.datatypes import GlobalType, MemoryType, Mutability, TableType, TypeIdx, ValType, addresses
+from wasm.datatypes.element_segment import ElemModeActive, ElemModeDeclarative, ElemModePassive
+from wasm.opcodes import BinaryOpcode
+from wasm.parsers import parse_module
+
+from pykwasm import kwasm_ast as a
+from pyk.kast.manip import split_config_from
+from pyk.kast.inner import KSequence, KSort, KToken, Subst
+from pyk.kore import syntax
+from pyk.ktool.krun import KRun
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+    from typing import IO
+
+    from pyk.kast import KInner
+    from wasm.datatypes import (
+        DataSegment,
+        ElementSegment,
+        Export,
+        Function,
+        FunctionType,
+        Global,
+        Import,
+        Limits,
+        Memory,
+        Module,
+        RefType,
+        StartFunction,
+        Table,
+    )
+    from wasm.datatypes.element_segment import ElemMode
+    from wasm.instructions import BaseInstruction
+
+
+def main():
+    # read env vars
+    DEBUG = 'DEBUG' in os.environ
+
+    # check arg count
+    if len(sys.argv) < 3:
+        print("usage: [DEBUG=1] run_wasm <llvm_dir> <wasm_file> [-cellname:sort=cellvalue...]")
+        sys.exit(1)
+    args = sys.argv[1:]
+
+    # parse fixed args
+    llvm_dir = Path(args[0])
+    wasm_file = Path(args[1])
+    infile = open(wasm_file, 'rb')
+
+    def build_subst_key(key_name):
+        return key_name.upper() + '_CELL'
+
+    # parse extra args
+    config_subst = dict()
+    extra_args = args[2:]
+    for arg in extra_args:
+        prekey_sort, val = arg[1:].split('=')
+        prekey, sort = prekey_sort.split(':')
+        key = build_subst_key(prekey)
+
+        if key == 'k':
+            raise ValueError(f"substitution may not contain a 'k' key")
+        if key in config_subst:
+            raise ValueError(f"redundant key found in substitution map: {prekey}")
+
+        if sort == 'String': val = f'"{val}"'
+        config_subst[key] = KToken(val, sort)
+
+    # parse module as binary (with fallback to textual parser)
+    try:
+        module = wasm2kast(infile)
+    except:
+        proc_res = subprocess.run(['wat2wasm', wasm_file, '--output=/dev/stdout'], check=True, capture_output=True)
+        infile.close()
+        infile = BytesIO(proc_res.stdout)
+        module = wasm2kast(infile)
+        infile.close()
+
+    # get runner
+    runner = KRun(llvm_dir)
+
+    # embed parsed_module to <k>
+    top_sort = KSort('GeneratedTopCell')
+    config_kast = runner.definition.init_config(top_sort)
+    symbolic_config, init_subst = split_config_from(config_kast)
+    init_subst['K_CELL'] = KSequence(module)
```

Comment:
In case you will wrap the module in a constructor in the other PR, you should also wrap it here.

Reply:
Thanks for this pointer.